network/tcp_client.cc

```diff
@@ -0,0 +1,490 @@
+/*
+Copyright 2025 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <chrono>
+
+#include "tcp_client.h"
+
+constexpr int      kKeepAliveIntervalSec = 5;
+constexpr int      kPingTimeoutMs = 5000;
+constexpr uint32_t kHandshakeMajorVersion = 1;
+constexpr uint32_t kHandshakeMinorVersion = 0;
+constexpr char     kPingMessage[] = "ping";
+constexpr char     kPongMessage[] = "pong";
+
+namespace Network
+{
+
+TcpClient::TcpClient() :
+    m_status(ClientStatus::DISCONNECTED)
+{
+    m_keep_alive.running = false;
+    m_keep_alive.interval_sec = kKeepAliveIntervalSec;
+    m_keep_alive.ping_timeout_ms = kPingTimeoutMs;
+}
+
+TcpClient::~TcpClient()
+{
+    Disconnect();
+}
+
+absl::Status TcpClient::Connect(const std::string& host, int port)
+{
+    if (GetClientStatus() == ClientStatus::CONNECTED ||
+        GetClientStatus() == ClientStatus::CONNECTING)
+    {
+        return absl::AlreadyExistsError(
+        "Connect: Client is already connected or in the process of connecting.");
+    }
+
+    StopKeepAlive();
+    if (m_connection)
+    {
+        m_connection->Close();
+        m_connection.reset();
+    }
+
+    SetClientStatus(ClientStatus::CONNECTING);
+    auto connection_or = SocketConnection::Create();
+    if (!connection_or.ok())
+    {
+        SetClientStatus(ClientStatus::STATUS_ERROR);
+        return absl::Status(connection_or.status().code(),
+                            absl::StrCat("Connect: ", connection_or.status().message()));
+    }
+
+    m_connection = std::move(connection_or.value());
+    auto conn_status = m_connection->Connect(host, port);
+    if (!conn_status.ok())
+    {
+        SetClientStatus(ClientStatus::STATUS_ERROR);
+        m_connection.reset();
+        return absl::Status(conn_status.code(),
+                            absl::StrCat("Connect: Connect fail: ", conn_status.message()));
+    }
+    SetClientStatus(ClientStatus::CONNECTED);
+
+    std::cout << "Client: Connected & handshaking." << std::endl;
+    auto handshake_status = PerformHandshake();
+    if (!handshake_status.ok())
+    {
+        SetClientStatus(ClientStatus::STATUS_ERROR);
+        if (m_connection)
+        {
+            m_connection->Close();
```

Comment:
it seems that m_connection->Close(); always come with m_connection.reset();, maybe a helper func to group those 2 calls?